# Activity 2: K-nn classification with credit data

#### Loading required libraries

In [1]:
library(dplyr)
library(class)


Attaching package: ‘dplyr’

The following objects are masked from ‘package:stats’:

    filter, lag

The following objects are masked from ‘package:base’:

    intersect, setdiff, setequal, union



#### Reading data from CSV file

In [2]:
data <- read.csv("german_credit_data1.csv")

#### Cleaning and reformating data

In [3]:
data <- na.omit(data)
levels(data$Sex) <- c('1', '2')
data$Sex <- as.numeric(as.character(data$Sex))
levels(data$Housing) <- c('1', '2', '3')
data$Housing <- as.numeric(as.character(data$Housing))
levels(data$Saving.accounts) <- c('1', '2', '3','4')
data$Saving.accounts <- as.numeric(as.character(data$Saving.accounts))
levels(data$Checking.account) <- c('1', '2', '3','4')
data$Checking.account <- as.numeric(as.character(data$Checking.account))
levels(data$Purpose) <- c('1', '2', '3','4','5','6','7','8')
data$Purpose <- as.numeric(as.character(data$Purpose))
data <- subset(data, select=c(2:11))
data

Age Sex Job Housing Saving.accounts Checking.account Credit.amount
2    22  1   2   2       1               2                 5951        
4    45  2   2   1       1               1                 7882        
5    53  2   2   1       1               1                 4870        
8    35  2   3   3       1               2                 6948        
10   28  2   3   2       1               2                 5234        
11   25  1   2   3       1               2                 1295        
12   24  1   2   3       1               1                 4308        
13   22  1   2   2       1               2                 1567        
14   60  2   1   2       1               1                 1199        
15   28  1   2   3       1               1                 1403        
16   32  1   1   2       2               1                 1282        
19   44  1   3   1       1               2                12579        
22   44  2   2   3       3               1                 2647        
23   48  2   1   3       1               1                 2241        
24   44  2   2   2       2               2                 1804        
26   36  2   1   2       1               1                 1374        
28   42  1   2   3       4               3                  409        
29   34  2   2   2       1               2                 2415        
30   63  2   2   2       1               1                 6836        
31   36  2   2   2       4               2                 1913        
32   27  2   2   2       1               1                 4020        
33   30  2   2   2       2               2                 5866        
35   33  1   3   2       1               3                 1474        
36   25  2   1   2       1               2                 4746        
38   37  2   2   2       1               3                 2100        
39   37  2   2   2       1               3                 1225        
40   24  2   2   2       1               2                  458        
42   26  2   2   2       3               2                 1158        
43   44  2   1   2       1               2                 6204        
44   24  2   2   3       2               1                 6187        
⋮    ⋮   ⋮   ⋮   ⋮       ⋮               ⋮                ⋮            
947  30  2   2   1       3               1                3349         
951  40  2   0   2       1               2                3590         
952  24  2   2   2       1               1                2145         
953  28  1   2   3       3               2                4113         
955  29  1   2   2       1               1                1893         
956  57  1   3   3       4               1                1231         
958  37  2   1   2       1               2                1154         
959  45  2   1   2       1               1                4006         
960  30  2   2   1       2               2                3069         
962  47  2   2   2       1               2                2353         
965  22  2   1   2       1               2                 454         
967  23  2   1   2       3               2                2520         
970  40  2   1   2       1               1                3939         
971  22  2   2   2       2               2                1514         
973  29  1   0   3       1               1                1193         
974  36  2   2   3       1               1                7297         
976  57  1   1   2       3               3                1258         
977  64  1   2   2       1               2                 753         
980  25  2   2   3       2               2                1264         
981  49  2   2   2       1               2                8386         
983  28  1   3   2       2               3                2923         
984  26  2   2   2       1               1                8229         
986  25  1   2   3       1               1                1433         
987  33  2   2   2       1               3   

#### Scaling data

In [4]:
new_data<-data.frame((scale(data)))

<b>K-nn process for :<br>Test Set : 20% <br>
Training Set : 80%</b>

In [5]:
# number of rows in the dataset
n.points <- nrow(new_data)
sampling.rate <- 0.8

#number of points in the test set
num.test.set.labels = n.points*(1-sampling.rate)

# randomly sample which rows will go in the training set
training <- sample(1:n.points, sampling.rate * n.points, replace=FALSE)

# define the training set to be those rows
train <- subset(new_data[training, ], select = c(1:9))

# the other rows are going into the test set
testing <- setdiff(1:n.points, training)

# define the test set to be the other rows
test <- subset(new_data[testing, ], select = c(1:9))

cl <- new_data$Credit.Risks[training]

# this is the subset of labels for the training set
true.labels <- new_data$Credit.Risks[testing]

In [6]:
#Loop through and calculate mismisclassification rate for different values of k
knn_misclassification1 <- NULL

for (k in 1:20) {
 predicted.labels <- knn(train, test, cl, k)
 num.incorrect.labels <- sum(predicted.labels != true.labels)
 misclassification.rate <- num.incorrect.labels / num.test.set.labels
 knn_misclassification1 <- rbind(knn_misclassification1, c(k, misclassification.rate))
}

#### Misclassification rate for values of k 1-20

In [7]:
colnames(knn_misclassification1)<-c("K-value","Misclassification rate")
knn_misclassification1
summary(knn_misclassification1)

K-value Misclassification rate
 [1,]  1      0.4022989             
 [2,]  2      0.3544061             
 [3,]  3      0.2681992             
 [4,]  4      0.3448276             
 [5,]  5      0.2969349             
 [6,]  6      0.3160920             
 [7,]  7      0.3352490             
 [8,]  8      0.3544061             
 [9,]  9      0.3256705             
[10,] 10      0.2969349             
[11,] 11      0.3160920             
[12,] 12      0.3352490             
[13,] 13      0.2873563             
[14,] 14      0.2873563             
[15,] 15      0.2873563             
[16,] 16      0.2490421             
[17,] 17      0.2298851             
[18,] 18      0.2873563             
[19,] 19      0.2586207             
[20,] 20      0.2777778

    K-value      Misclassification rate
 Min.   : 1.00   Min.   :0.2299        
 1st Qu.: 5.75   1st Qu.:0.2850        
 Median :10.50   Median :0.2969        
 Mean   :10.50   Mean   :0.3056        
 3rd Qu.:15.25   3rd Qu.:0.3352        
 Max.   :20.00   Max.   :0.4023        

<b> We can observe from this data set that the minimum misclassification rate for training data : 80% and test data: 20% is 0.2299</b>

<b>K-nn misclassification rate for :<br>Test Set : 30% <br>
Training Set : 70%</b>

In [8]:
sampling.rate <- 0.7

#number of points in the test set
num.test.set.labels = n.points*(1-sampling.rate)

# randomly sample which rows will go in the training set
training <- sample(1:n.points, sampling.rate * n.points, replace=FALSE)

# define the training set to be those rows
train <- subset(new_data[training, ], select = c(1:9))

# the other rows are going into the test set
testing <- setdiff(1:n.points, training)

# define the test set to be the other rows
test <- subset(new_data[testing, ], select = c(1:9))

cl <- new_data$Credit.Risks[training]

# this is the subset of labels for the training set
true.labels <- new_data$Credit.Risks[testing]

In [9]:
#Loop through and calculate misclassification rate for different values of k
knn_misclassification2 <- NULL

for (k in 1:20) {
 predicted.labels <- knn(train, test, cl, k)
 num.incorrect.labels <- sum(predicted.labels != true.labels)
 misclassification.rate <- num.incorrect.labels / num.test.set.labels
 knn_misclassification2 <- rbind(knn_misclassification2, c(k, misclassification.rate))
}

#### Misclassification rate for values of k 1-20

In [10]:
colnames(knn_misclassification2)<-c("K-value","Misclassification rate")
knn_misclassification2
summary(knn_misclassification2)

K-value Misclassification rate
 [1,]  1      0.3831418             
 [2,]  2      0.3959132             
 [3,]  3      0.3384419             
 [4,]  4      0.3384419             
 [5,]  5      0.3512133             
 [6,]  6      0.3767561             
 [7,]  7      0.3512133             
 [8,]  8      0.3575990             
 [9,]  9      0.3703704             
[10,] 10      0.3767561             
[11,] 11      0.3639847             
[12,] 12      0.3320562             
[13,] 13      0.3512133             
[14,] 14      0.3320562             
[15,] 15      0.3512133             
[16,] 16      0.3575990             
[17,] 17      0.3575990             
[18,] 18      0.3512133             
[19,] 19      0.3639847             
[20,] 20      0.3831418

    K-value      Misclassification rate
 Min.   : 1.00   Min.   :0.3321        
 1st Qu.: 5.75   1st Qu.:0.3512        
 Median :10.50   Median :0.3576        
 Mean   :10.50   Mean   :0.3592        
 3rd Qu.:15.25   3rd Qu.:0.3720        
 Max.   :20.00   Max.   :0.3959        

<b>We can observe from this data set that the minimum misclassification rate for training data : 70% and test data: 30% is 0.3321</b>